rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,190 @@
+connection, but using a single connection for a large number of streams is 
+problematic for a few reasons:
+
+- Contention of the upstream connection.
+- TCP head of line blocking becomes more problematic as the amount of data being
+  transferred over a single connection increases.
+- The server enforces a maximum number of streams per connection.
+
+To avoid this, a simple connection cycler will be implemented. After a number of
+streams have been established, a new internal connection will be dialed. New
+streams will use the new connection, and the old connection will be closed once
+all streams using it have been closed.
+
+### Initial Scope
+
+For the initial version, we must include:
+
+- The long-lived proxying background service within `tbot`
+  - This will include a rudimentary connection cycler.
+- The new lightweight client binary with support for ProxyUseFdPass.
+- A configuration option to control the maximum number of proxied connections.
+  - When this maximum is reached, the service will wait until the number of
+    active connections drops below the maximum.
+  - This will allow a sane limit to be provided to prevent the long-lived `tbot`
+    process from being overwhelmed.
+- Prometheus metrics which exposes the health of the proxying.
+  - Number of active proxied connections.
+  - Number of attempts to establish proxied connections, the time this has
+    taken, and the success status.
+- OpenTelemetry trace spans at key points.
+  - This will allow us to target future performance improvements.
+
+### Future Scope
+
+#### Caching
+
+We can further improve performance by caching the results of operations such as
+resolving the target host from labels.
+
+This work is deferred as it is complex and introduces risks. Information from
+OpenTelemetry tracing will allow us to target the most impactful areas at a 
+later date.
+
+#### Further improving re-use of connections
+
+The initial version will use a rudimentary connection "cycler". This will cause
+a new connection to be established after a certain number of downstream
+connections have been established. It will then close the upstream connection
+once all downstream connections using it have been closed.
+
+This could be further improved by balancing incoming downstream connections
+against a pool of upstream connections, because the upstream connections would
+live longer and we'd avoid a scenario where an upstream connection remains open
+but is only being utilised by one or two remaining downstream connections.
+
+### UX
+
+The user will configure `ssh-local-proxy` in the `tbot` configuration file:
+
+```yaml
+services:
+  - type: ssh-local-proxy
+    # Destination is where the ssh_config and UDS will be written to. It must be
+    # a directory.
+    destination:
+      type: directory
+      path: /opt/machine-id
+    # The maximum number of connections that can be proxied at once.
+    max_connections: 100
```

Comment:
Is this the limit per grpc connection, or global for the whole service?

Reply:
[9866425](https://github.com/gravitational/teleport/pull/42546/commits/98664250527be2ab02dd8b914e25d4ab9486b195)